### PR TITLE
Improve semchecking for duplicate cases in case statements

### LINKED ISSUE #6239: Duplicate case values because of duplicates in set
The following code results in a c/c++ compiler crash because nim creates duplicate case values for `3`:

```nim
const x = { 1, 2, 3, 3 }
let y = 1

case y:
of x: discard
else: discard
```

Perhaps nim could also give a warning when there are duplicate values in set literals?

### PR DESCRIPTION
**Pull request by GULPF**
The previous code had an issue where only the first element of an array/set where checked for duplicates. Since set constructors are allowed to have duplicates, I've also made sure that sets in case statements are deduplicated before they are semchecked.

Fixes #6239

### CODE REVIEW

**Comment by Araq** [comment]
I'm not really happy with this solution. Why is it required to change the AST in order to properly count these things?

**Comment by GULPF** [comment]
What AST change are you referring to? The only AST transformation that this PR introduces is the deduplication, which I think is fine since the codegen must do it otherwise.

**Comment by cooldome** [comment]
I'll try to explain, maybe Araq can do it better and correct me: semcheck pass is supposed not to change the AST unless it is impossible to achieve the solution any other way, then there is transformation pass (transf) that can and should deduplicate

**Comment by GULPF** [comment]
Thanks for the explanation, that makes sense. Should I just move the deduplication, or should I refactor `semCaseBranch` so that all the structural transformations are done in transf.nim? ping @Araq

**Comment by Araq** [comment]
Remove the deduplication, leave the rest as it is.

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -455,13 +455,15 @@ proc semIdentWithPragma(c: PContext, kind: TSymKind, n: PNode,
...
   if gCmd == cmdPretty: styleCheckDef(n.info, result)
 
-proc checkForOverlap(c: PContext, t: PNode, currentEx, branchIndex: int) =
-  let ex = t[branchIndex][currentEx].skipConv
-  for i in countup(1, branchIndex):
-    for j in countup(0, sonsLen(t.sons[i]) - 2):
-      if i == branchIndex and j == currentEx: break
-      if overlap(t.sons[i].sons[j].skipConv, ex):
-        localError(ex.info, errDuplicateCaseLabel)
+proc checkForOverlap(seenCases: var seq[PNode], n: PNode,
```
Comment (Araq) [inline compiler/semtypes.nim]: Can 'seenCases' be an IntSet instead?

**Comment by Araq** [comment]
Oh sorry, this solution looks worse than the one before. Hmm. Revert it and live with the transformation in the sem pass?

**Comment by GULPF** [comment]
Agreed, both deduplication and semchecking became harder after I split the transformations to different places. I think the `case` transformations must happen at the same place.

File: tests/casestmt/tduplicates.nim [hunk 2]
```diff
@@ -0,0 +1,47 @@
+discard """
+  action: run
```
Comment (Araq) [inline tests/casestmt/tduplicates.nim]: This should have some output and use the 'output' section.

**Pull request merged**